lib/floe/runner/container_runner.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require_relative "container_runner/docker_mixin"
+require_relative "container_runner/docker"
+require_relative "container_runner/kubernetes"
+require_relative "container_runner/podman"
+
+module Floe
+  class Runner
+    class ContainerRunner
+      class << self
+        def cli_options(optimist)
+          optimist.banner("")
+          optimist.banner("Container runner options:")
+
+          optimist.opt :container_runner, "Type of runner for docker container images (docker, podman, or kubernetes)", :type => :string, :short => 'r'
+          optimist.opt :container_runner_options, "Options to pass to the container runner", :type => :strings, :short => 'o'
+
+          optimist.opt :docker,     "Use docker to run container images     (short for --container-runner=docker)",     :type => :boolean
+          optimist.opt :podman,     "Use podman to run container images     (short for --container-runner=podman)",     :type => :boolean
+          optimist.opt :kubernetes, "Use kubernetes to run container images (short for --container-runner=kubernetes)", :type => :boolean
+        end
+
+        def resolve_cli_options!(opts)
+          # shortcut support
+          opts[:container_runner] ||= "docker" if opts[:docker]
+          opts[:container_runner] ||= "podman" if opts[:podman]
+          opts[:container_runner] ||= "kubernetes" if opts[:kubernetes]
+
+          runner_options = opts[:container_runner_options].to_h { |opt| opt.split("=", 2) }
+
+          begin
+            set_runner(opts[:container_runner], runner_options)
+          rescue ArgumentError => e
+            Optimist.die(:container_runner, e.message)
+          end
+        end
+
+        def runner
+          @runner || set_runner(nil)
+        end
+
+        def set_runner(name_or_instance, options = {})
+          @runner =
+            case name_or_instance
+            when "docker", nil
+              Floe::Runner::ContainerRunner::Docker.new(options)
+            when "podman"
+              Floe::Runner::ContainerRunner::Podman.new(options)
+            when "kubernetes"
+              Floe::Runner::ContainerRunner::Kubernetes.new(options)
+            when Floe::Workflow::Runner
+              name_or_instance
+            else
+              raise ArgumentError, "container runner must be one of: docker, podman, kubernetes"
+            end
+        end
+      end
+    end
+  end
+end
+
+Floe::Workflow::Runner.register_scheme("docker", -> { Floe::Runner::ContainerRunner.runner })
```

Comment:
I like the idea of also adding
```
Floe.register_plugin(Floe::Runner::ContainerRunner.new)
```

yes, the methods don't need to be an instance, but makes it easier to be called.
Hmm. this will work if we just pass in the class, since it will call the method on anything passed in.

Also like register_scheme to be called from `set_runner` - but that is a tomato/potatoe kind of thing. (agree to disagree and not my hill)

Reply:
I considered something like that but it just felt like another level of unnecessary indirection, and this made it seems a lot more explicit what was happening.  This is something we can do in a followup because it's the plugin's API and there's only 1 plugin at the moment, so it's a an easy change if that's what we want.